src/agent.c

```diff
@@ -2651,12 +2713,40 @@ void agent_update_ordered_pairs(juice_agent_t *agent) {
 	}
 }
 
-static inline bool pair_is_relayed(const ice_candidate_pair_t *pair) {
-	return pair->local && pair->local->type == ICE_CANDIDATE_TYPE_RELAYED;
-}
+bool agent_address_uses_tcp(juice_agent_t *agent, const addr_record_t *addr) {
+	bool result = false;
+	conn_lock(agent);
+	agent_stun_entry_t *selected_entry = atomic_load(&agent->selected_entry);
+	if (selected_entry && selected_entry->pair && selected_entry->pair->remote) {
+		bool compare_ports = entry_compare_ports(selected_entry);
+		bool match = addr_record_is_equal(&selected_entry->record, addr, compare_ports);
+		if (!match)
+			match = addr_record_is_equal(&selected_entry->pair->remote->resolved, addr, compare_ports);
+		if (match) {
+			result = selected_entry->pair->remote->transport != ICE_CANDIDATE_TRANSPORT_UDP;
+			conn_unlock(agent);
+			return result;
+		}
+	}
 
-static inline bool entry_is_relayed(const agent_stun_entry_t *entry) {
-	return entry->pair && pair_is_relayed(entry->pair);
+	for (int i = 0; i < agent->entries_count; ++i) {
+		agent_stun_entry_t *entry = agent->entries + i;
+		if (entry->type != AGENT_STUN_ENTRY_TYPE_CHECK || !entry->pair || !entry->pair->remote)
+			continue;
+		bool compare_ports = entry_compare_ports(entry);
+		bool match = addr_record_is_equal(&entry->record, addr, compare_ports);
+		if (!match)
+			match = addr_record_is_equal(&entry->pair->remote->resolved, addr, compare_ports);
+		if (!match)
+			continue;
+		if (entry->pair->remote->transport != ICE_CANDIDATE_TRANSPORT_UDP) {
+			result = true;
+			break;
+		}
+	}
+
+	conn_unlock(agent);
+	return result;
 }
```

Comment:
Don't do this. If the address record has `socktype == SOCK_STREAM`, it's TCP, otherwise it's UDP. It's as simple as that.